src/models/geograph.py

```diff
@@ -443,27 +450,143 @@ def merge_nodes(
+        max_travel_distance: float = 0.0,
+    ) -> None:
+        """
+        Create habitat graph and store it in habitats dictionary.
+
+        Creates a habitat subgraph of the main GeoGraph that only contains edges
+        between nodes in `valid_classes` as long as they are less than
+        `max_travel_distance` apart. All nodes which are not in `valid_classes`
+        are not in the resulting habitat graph.
+
+        Args:
+            name (str): The name of the habitat.
+            valid_classes (set): A set of class labels which make up the habitat.
+            max_travel_distance (float): The maximum distance the animal(s) in
+            the habitat can travel through non-habitat areas. The habitat graph
+            will contain edges between any two nodes that have a class label in
+            `valid_classes`, as long as they are less than `max_travel_distance`
+            units apart. Defaults to 0, which will only create edges between
+            directly neighbouring areas.
+        """
+        hgraph: nx.Graph = deepcopy(self.graph)
+        hgraph.graph["name"] = name
+        # Remove all edges in the graph, then at the end we only have edges
+        # between nodes less than `max_travel_distance` apart
+        hgraph.clear_edges()
+        # Vectorised buffer on the entire df to calculate the expanded polygons
+        # used to get intersections.
+        self.df["buff_polygon"] = self.df.geometry.buffer(max_travel_distance)
+        # List of nodes that are not in the habitat
+        invalid_indexes: List[int] = []
+        for node in tqdm(
+            hgraph.nodes, desc="Generating habitat graph", total=len(hgraph)
+        ):
+            if hgraph.nodes[node]["class_label"] not in valid_classes:
+                invalid_indexes.append(node)
+                continue
+            # df.at is a fast df.loc designed only to access single values
+            polygon = self.df.at[node, "geometry"]
+            buff_poly = self.df.at[node, "buff_polygon"]
+            # Query rtree for all polygons within `max_travel_distance` of the original
+            poss_neighbours = list(self.rtree.intersection(buff_poly.bounds))
+            for nbr in poss_neighbours:
+                # If a node is not a habitat class node, don't add the edge
+                if hgraph.nodes[nbr]["class_label"] not in valid_classes:
+                    continue
+                # Otherwise add the edge with distance attribute
+                nbr_polygon = self.df.at[nbr, "geometry"]
+                if not hgraph.has_edge(node, nbr) and nbr_polygon.intersects(buff_poly):
+                    # Since intersects is cheaper than distance, we can save time here
+                    if nbr_polygon.intersects(polygon):
+                        distance = 0.0
+                    else:
+                        distance = polygon.distance(nbr_polygon)
+                    hgraph.add_edge(node, nbr, distance=distance)
```

Comment:
Great that you're computing distance as well, that will be useful for metrics! We might wan't to make the distance computation optional though, depending on how computationally expensive it is compared to the rest. The idea would be that users might only want to quickly visualise a habitat without metrics. Do you think that would potentially save a lot of time for users?

Reply:
I tested it for some large habitats and it does shave off a decent amount of time, so I'll make it optional. If it is a very large habitat though it will still take ~20-40 seconds to load, but small habitats with only a few thousand nodes are usually less than 10s.

Reply:
Great, thanks for testing it!